Alonzo support

davidmisiak opened the pull request:
Changes:
- Added support for signing txs with Alonzo-related items.
- Added support for the new CDDL-compliant tx format.
- Marked the cardano-cli's internal raw tx format deprecated; marked the `transaction sign` call deprecated (in favor of `transaction witness`).
- Upgraded Node.js to v14.

TODO:
- [ ] add a test case for `includeNetworkId: true`

davidmisiak commented on src/crypto-providers/util.ts:
```diff
@@ -199,37 +199,26 @@ const filterSigningFiles = (
-const findPathForKeyHash = (
-  keyHash: Buffer, signingFiles: HwSigningData[],
-): BIP32Path | undefined => {
-  const signingFile = signingFiles.find((file) => keyHash.equals(hwSigningFileToPubKeyHash(file)))
-  return signingFile?.path
+const hwSigningFileToPubKeyHash = (signingFile: HwSigningData): Uint8Array => {
+  const pubKey = hwSigningFileToPubKey(signingFile)
+  return cardanoSerialization.PublicKey.from_bytes(pubKey).hash().to_bytes()
 }
 
-const findSigningPathForKeyHash = (
```
These two functions seem to behave exactly the same. I discussed with Jakub why he added `findPathForKeyHash` while keeping `findSigningPathForKeyHash` (the original one) as well, but he said he doesn't remember and recommended keeping only one of them. I chose the one which uses `cardanoSerialization` since we consider it generally more up-to-date than `cardanoCrypto` if I understand correctly (I remember something like this from previous conversations, please correct me if this is wrong).

janmazak replied:
That sounds good.

davidmisiak commented on src/crypto-providers/ledgerCryptoProvider.ts:
```diff
@@ -455,6 +458,38 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
+      }
+  }
+
   const prepareAdditionalWitnessRequests = (
     mintSigningFiles: HwSigningData[],
     multisigSigningFiles: HwSigningData[],
```
How should we modify `prepareAdditionalWitnessRequests` for Alonzo? In Alonzo txs, successful Plutus script validation may require eg. three 1852 signatures, but if the tx contains only one input and one collateral input, there is no way how to force hw-cli to also create the third 1852 signature (except for calling `witness` call manually). Ideally, the signers required by Plutus script should be listed in `requiredSigners`, but I'm afraid we cannot rely on this always being the case.

The best solution I can think of right now is to modify it like this:
```ts
let files = [...mintSigningFiles ...multisigSigningFiles]
if (signingMode === SigningMode.PLUTUS_TRANSACTION) {
    files = [...files, ...paymentSigningFiles, ...stakeSigningFiles] 
}
return files.map((f) => f.path)
```

This way, some unwanted 1852 signatures will be created if the user provides superfluous signing files, but this is limited only to the Plutus signing mode (where the user sees all witness requests on the hw wallet screen). Furthermore, we are already doing this for 1854 signing files in multisig signing mode.

@janmazak what do you think?

janmazak replied:
If superfluous witnesses appear only if the user explicitly supplies some signing files, I see no real problem, it's a bit of an inconvenience at most. It might be resolved in the future if it bothers someone.

Perhaps add a comment explaining the situation.

davidmisiak replied:
Dex guys notified us that Plutus scripts can not access the signatures - they see only required signers (and the other tx items). Thanks to this, I believe we don't have to modify `prepareAdditionalWitnessRequests`. However, I will add a comment explaining this, in order not to confuse someone in the future.

davidmisiak commented:
I noticed that [`ensureFirmwareSupportsParams` function](https://github.com/vacuumlabs/cardano-hw-cli/blob/develop/src/crypto-providers/trezorCryptoProvider.ts#L380-L402) wasn't updated for multisig. However, I believe that the exact same checking is performed by Connect and Ledgerjs, at the beginning of the signing call, so this seems redundant to me and I suggest removing it, together with the [feature definitions](https://github.com/vacuumlabs/cardano-hw-cli/blob/develop/src/crypto-providers/constants.ts).

For reference, this functionality was added in [here](https://github.com/vacuumlabs/cardano-hw-cli/pull/42#discussion_r568808631). I haven't found any specific reasoning why it was added.

janmazak commented:
We would only need a check like `ensureFirmwareSupportsParams` if ledgerjs/connect were not capable of reporting an error. In particular, if ledgerjs did not contain a function hw-cli is going to call at all. But since hw-cli is never updated beyond ledgerjs/connect, I don't see any need for that.

davidmisiak commented on src/commandExecutor.ts:
```diff
@@ -98,19 +99,39 @@ const CommandExecutor = async () => {
+      tx = InteropLib.decodeTx(txCbor)
+    }
+
+    const txBody = (rawTx?.body ?? tx?.body)!
```
This is rather ugly, but I lean towards doing it this way instead of creating a proper union type, since we will remove `rawTxFileData` altogether in a few months.

janmazak commented on src/crypto-providers/ledgerCryptoProvider.ts:
```diff
@@ -580,12 +602,23 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
+    const scriptDataHashHex = prepareScriptDataHash(body.scriptDataHash)
+    const collaterals = body.collaterals?.map(
+      (collateralInput, i) => prepareCollateralInput(
+        collateralInput, getSigningPath(paymentSigningFiles, inputs.length + i),
```
This deserves an explanation in a comment. Why is `inputs` used here at all when we are dealing with collateral?

janmazak replied:
The role of the function `getSigningPath` is unclear to me. Can we get rid of it, or refactor/rename/replace it somehow in a way that will make obvious what's going on?

davidmisiak replied:
Comments added in https://github.com/vacuumlabs/cardano-hw-cli/pull/112/commits/39dca2c9884367933e86f6f1109495f4302ad9f0. I couldn't think of a simple and clear refactor, but I believe that the explanation in `getSigningPath` is pretty clear.

janmazak replied:
If I understand correctly what is going on, it all seems very Potemkin-like... and indicates some potential problems with the APIs.

As far as I can see, ledgerjs does not require having a witness path attached to every input. (I've not checked about Trezor, but since multisig there are inputs possibly signed by several paths anyway, so I don't think there should be a requirement of having one of them directly attached to the input.)

Hence a more natural way for cardano-hw-cli would be to just pass all payment signing file paths in `additionalWitnessPaths`, even those for inputs. It truly does not know which path belongs to each input, so I guess it should not be trying to fake that knowledge.

davidmisiak replied:
As discussed offline, we will resolve this separately. I created an issue for this: https://github.com/vacuumlabs/cardano-hw-cli/issues/120

janmazak reviewed (commented):
Not finished (I got to `ledgerTypes.ts` which I have not read yet).

janmazak commented on src/crypto-providers/util.ts:
```diff
@@ -199,37 +198,26 @@ const filterSigningFiles = (
-): BIP32Path | undefined => {
-  const signingFile = signingFiles.find((file) => keyHash.equals(hwSigningFileToPubKeyHash(file)))
-  return signingFile?.path
+const hwSigningFileToPubKeyHash = (signingFile: HwSigningData): Uint8Array => {
```
Why are we using `Uint8Array` here and `Buffer` in most other places?

davidmisiak replied:
Probably because cardano-serialization-lib's `to_bytes()` returns `Uint8Array` (btw, `Buffer` is a proper subclass of `Uint8Array`). I changed `hwSigningFileToPubKeyHash` to return `Buffer`.

janmazak commented on src/crypto-providers/util.ts:
```diff
@@ -267,6 +255,10 @@ const determineSigningMode = (
       : SigningMode.POOL_REGISTRATION_AS_OWNER
   }
 
+  if (txBody.collaterals || txBody.requiredSigners) {
```
`determineSigningMode` is very fragile. At the least it requires detailed comments explaining why it is written exactly like this. For instance, one of my immediate ideas was to move the multisig signing files check above txBody checks (because it seems like an explicitly expressed intent --- which it actually is not, since we allow 1854 within Plutus).

davidmisiak replied:
Good point, comments added in https://github.com/vacuumlabs/cardano-hw-cli/pull/112/commits/760d83b1949aa852608df595426094d89e8f70f1

janmazak commented on src/transaction/transactionValidation.ts:
```diff
@@ -106,12 +119,13 @@ const transformTx = (args: ParsedTransactionTransformArguments): void => {
     console.log('Fixed transaction will be written to the output file.')
```
I don't think "fix" is specific enough. Perhaps "transformed"? (in some other place, too)

davidmisiak replied:
I changed this to "transformed". We use the word "fix" when describing "fixable/unfixable" errors during validation, but it cannot be changed so easily and we mention running the "transform" command there.

janmazak reviewed (commented):
Still not finished (I'll continue with `transaction.ts`).

janmazak commented on src/transaction/transaction.ts:
```diff
@@ -1,45 +1,128 @@
+  if (nativeScriptWitnessList.length > 0) {
+    witnessSet.set(TxWitnessKeys.NATIVE_SCRIPTS, nativeScriptWitnessList)
   }
+
```
I guess I'd move this up near shelleyWitnesses, that's where it semantically belongs and how treat them in the code (irrespective of what the enum values are).

davidmisiak replied:
Moved

janmazak commented on src/transaction/transaction.ts:
```diff
@@ -1,45 +1,128 @@
+    return _rawTxToTxSigned(params, byronWitnesses, shelleyWitnesses)
+  }
+
+  const witnessSet = _parseWitnessSet(params.tx!.witnessSet)
```
I'd welcome a comment explaining we only need to add witnesses provided by the HW wallet and everything else (e.g. scripts) stays the same.

davidmisiak replied:
Added

janmazak commented on src/types.ts:
```diff
@@ -101,18 +102,22 @@ export enum NetworkIds {
   MAINNET = 764824073,
-  TESTNET = 42,
+  TESTNET = 1097911063,
+  TESTNET_LEGACY = 42,
```
This deserves an explanation.

davidmisiak replied:
Added in https://github.com/vacuumlabs/cardano-hw-cli/pull/112/commits/3680f30c7cc94d89844d339b08d1d564039d81cc

janmazak reviewed (commented):
Finished for now.

janmazak commented:
ledgerjs needs an update to rc5 at least

janmazak commented:
`yarn audit` shows lots of problems

davidmisiak commented on docs/catalyst-voting-registration-example.md:
````diff
@@ -48,28 +48,36 @@ cardano-hw-cli catalyst voting-key-registration-metadata \
-Create payment hwsfile with `cardano-hw-cli`:
+HW wallets expect the transaction CBOR to be in *canonical* format. Unfortunately, cardano-cli sometimes produces incorrectly formatted tx files. Use the following command to fix the formatting issues.
 ```
-cardano-hw-cli address key-gen \
---path 1852H/1815H/0H/0/0 \
---verification-key-file payment.vkey \
---hw-signing-file payment.hwsfile
````
I guess I removed this because it doesn't really make sense here - we've already built the tx (with some inputs) and we are only now generating the signing file to sign it? I will at least add a mention that `payment.hwsfile` is the one we used when building the tx.

davidmisiak commented on docs/token-minting.md:
````diff
@@ -58,7 +58,11 @@ cardano-cli transaction policyid --script-file policy.script
+
+In this transaction we are going to mint `1000000` tokens named `C0in`. First, we need to encode the token name:
+```sh
+TOKEN_NAME=$(echo -n "C0in" | xxd -b -ps -c 80 | tr -d '\n')
````
I believe so - `cardano-cli` reported errors, and this step is present in the official `cardano-cli` docs.

gitmachtl commented:
hey @davidmisiak , some questions about:

* Added support for the new CDDL-compliant tx format.
* Marked the cardano-cli's internal raw tx format deprecated

With my current SPO scripts (using hw-cli 1.9.1) i am generating the txBodyFile without any additional parameter on cardano-cli.  the new cardano-cli 1.34.1 is able to generate the output cbor format also in CDDL format. I am a bit confused how the new hw-cli will handle both formats. So will it accept both formats until October 2022?

I have not looked at the changes to the "transaction transform-raw" -> "transaction transform" routines, but will the "transform" also work on both formats?

gitmachtl commented:
I did a test with a CDDL txbody input-file:
``` json
{
    "type": "Unwitnessed Tx MaryEra",
    "description": "Ledger Cddl Format",
    "cborHex": "83a500818258209307a53bc38a6a223e70d2e5650f60607a403b4efba2509cd3d9fcb731e87f4100018182583900897789a996ec52f93668d18dd4dbe5f00036a0406f6da37185c641caf695c35024868c4dd6f694e16b88bdf6986e31a074f790d75241c44b821b0000000f38242545a5581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a14548616c6c6f05581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a940181e43414441192328494154414441636f696e19277448486170707941444119138845546573743108455465737432144554657374330a4873636f6f70209a0a1903b74873636f6f70209b0a1903b7581c4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467a24500010203041903914505040302011903e8581c9c3f26285634740a52c31e5fcd0686fe51a0acb2dbcd9e565d96b736a140181b581cecd07b4ef62f37a68d145de8efd60c53d288dd5ffc641215120cc3dba1400f021a0002f1bd031a03307f3109a1581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a1494154414441636f696e1864a101818200581c45d70e54f3b5e9c5a2b0cd417028197bd6f5fa5378c2f5eba896678df6"
}
```

``` console
cardano-hw-cli transaction transform --tx-file /tmp/cddl-test.payment.txbody --out-file /dev/stdout

The transaction contains following fixable errors:
- CBOR is not canonical (transaction)
Fixed transaction will be written to the output file.
```
``` json
{
    "type": "Unwitnessed Tx MaryEra",
    "description": "",
    "cborHex": "83a500818258209307a53bc38a6a223e70d2e5650f60607a403b4efba2509cd3d9fcb731e87f4100018182583900897789a996ec52f93668d18dd4dbe5f00036a0406f6da37185c641caf695c35024868c4dd6f694e16b88bdf6986e31a074f790d75241c44b821b0000000f38242545a5581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a14548616c6c6f05581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a940181e4341444119232845546573743108455465737432144554657374330a4848617070794144411913884873636f6f70209a0a1903b74873636f6f70209b0a1903b7494154414441636f696e192774581c4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467a24500010203041903914505040302011903e8581c9c3f26285634740a52c31e5fcd0686fe51a0acb2dbcd9e565d96b736a140181b581cecd07b4ef62f37a68d145de8efd60c53d288dd5ffc641215120cc3dba1400f021a0002f1bd031a03307f3109a1581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a1494154414441636f696e1864a101818200581c45d70e54f3b5e9c5a2b0cd417028197bd6f5fa5378c2f5eba896678df6"
}
```

Is the outputfile again in cddl-format? If so it would be nice to have the description field copied over from the input file to the output file.

gitmachtl commented:
I further tried to use Cardano-App 4.0.0 (branch alonzo) to do a transaction, but it failed:

``` console
{
    "type": "Unwitnessed Tx MaryEra",
    "description": "Ledger Cddl Format",
    "cborHex": "83a400838258205c247b4b92eeaff45dd054f89f8a5b3487c9f3842ac34659b1d95a98fab9e3ae008258205dd2248a5d442e9a127edbc1b14396c5a71c89c47cb39cb1ca15e923771718af008258206535cae655341bff6c1c2d4ac5af877046fd270508d4eec76f708e93be04faa40001818258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d51c0474a3581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a44018644611223322110018c84568616c6c6f19012c43787878190190581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a243787878190190429999190190021a0002d981031a03308ccaa0f6"
}

Autocorrect the TxBody for canonical order: Fixed transaction will be written to the output file.

{
    "type": "Unwitnessed Tx MaryEra",
    "description": "",
    "cborHex": "83a400838258205c247b4b92eeaff45dd054f89f8a5b3487c9f3842ac34659b1d95a98fab9e3ae008258205dd2248a5d442e9a127edbc1b14396c5a71c89c47cb39cb1ca15e923771718af008258206535cae655341bff6c1c2d4ac5af877046fd270508d4eec76f708e93be04faa40001818258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d51c0474a3581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a4401864437878781901904568616c6c6f19012c4611223322110018c8581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a242999919019043787878190190021a0002d981031a03308ccaa0f6"
}

Sign the unsigned transaction body with the ledger-red.payment.hwsfile:  /tmp/ledger-red.payment.tx

Cardano App Version 4.0.0 found on your Ledger device!
Please approve the action on your Hardware-Wallet (abort with CTRL+C) ... Error: Invalid file contents of tx-body-file
```

davidmisiak commented:
> With my current SPO scripts (using hw-cli 1.9.1) i am generating the txBodyFile without any additional parameter on cardano-cli. the new cardano-cli 1.34.1 is able to generate the output cbor format also in CDDL format. I am a bit confused how the new hw-cli will handle both formats. So will it accept both formats until October 2022?

Yes, until October 2022, we will support both the cardano-cli's rawTx format and the new CDDL-compliant tx format. Then we will drop support for the first one, because now it's possible to generate CDDL-compliant transactions simply by adding `--cddl-format` to `cardano-cli transaction build[-raw]`.

> Is the outputfile again in cddl-format? If so it would be nice to have the description field copied over from the input file to the output file.

Good point, I believe we can copy the description when transforming the transaction.

> I further tried to use Cardano-App 4.0.0 (branch alonzo) to do a transaction, but it failed (cardano-hw-cli from dm/alonzo):
> Error: Invalid file contents of tx-body-file

I'm not sure I follow, could you maybe post the `cardano-cli` and `cardano-hw-cli` commands that the script calls please? Generally, this looks like a wrong combination of commands - if you use `cardano-cli transaction build[-raw]`, make sure you use `cardano-hw-cli` commands `validate-raw`, `transform-raw`, `transaction sign/witness --tx-body-file`; if you've added `--cddl-format`, you should use `cardano-hw-cli` commands `validate`, `transform`, `transaction sign/witness --tx-file`.

> Repeating the test with the --cli-format setting was ok:

The first example has `Unwitnessed Tx MaryEra` envelope type, which should be the new CDDL-compliant format, whereas the second one has `TxBodyMary`, which is the old rawTx format. Are you sure you used `--cddl-format` for the second example and not the other way around?

davidmisiak commented:
Ah yes, my bad, I misread --cli-format vs. --cddl-format. So yes, if the first example is in the CDDL format, then `Error: Invalid file contents of tx-body-file` woudl appear if you try to sign it with `cardano-hw-cli transaction sign --tx-body-file` instead of `--tx-file` - is there a chance this is the case?

gitmachtl commented:
> Ah yes, my bad, I misread --cli-format vs. --cddl-format. So yes, if the first example is in the CDDL format, then `Error: Invalid file contents of tx-body-file` woudl appear if you try to sign it with `cardano-hw-cli transaction sign --tx-body-file` instead of `--tx-file` - is there a chance this is the case?

Ahhh... yes, could be the case. Let me retest it.

gitmachtl commented:
> Ah yes, my bad, I misread --cli-format vs. --cddl-format. So yes, if the first example is in the CDDL format, then `Error: Invalid file contents of tx-body-file` woudl appear if you try to sign it with `cardano-hw-cli transaction sign --tx-body-file` instead of `--tx-file` - is there a chance this is the case?

yep, that was the case. so the cddl-format is working. i also tested it with an alonzoera transaction, that worked too.

but the output json fields of the signed tx file are a bit strange, it changes from the unsigned tx file:
``` json
{
    "type": "Unwitnessed Tx AlonzoEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a60081825820..."
}
```
to

``` json
{
    "type": "Unwitnessed Tx AlonzoEra",
    "description": "",
    "cborHex": "84a400818258206274306..."
}
```
after the signing. So seeing it here again as type `Unwitnessed Tx AlonzoEra` is a bit confusing.
Shouldn't it be `Witnessed Tx AlonzoEra` here? And, is the file at this stage also again in the CDDL format?

davidmisiak commented on src/commandExecutor.ts:
```diff
@@ -98,19 +99,39 @@ const CommandExecutor = async () => {
-      signingMode: determineSigningMode(rawTx.body, args.hwSigningFileData),
+      signingMode: determineSigningMode(txBody, args.hwSigningFileData),
       rawTx,
-      txBodyHashHex: getTxBodyHash(rawTx.body),
+      tx,
```
Yeah, as I mentioned above, this is quite bad, but we will be removing this in a few months anyway so I guess creating a proper union type and then undoing the changes is not worth it.

davidmisiak commented on src/command-parser/parsers.ts:
```diff
@@ -109,8 +114,12 @@ export const parseTxFile = (path: string): TxFileData => {
+    era,
+    envelopeType: type,
+  }
   if (isTxFileData(data)) {
```
Technically, it doesn't (the envelope type is validated by selecting the corresponding era), but I added the validation just to make sure. https://github.com/vacuumlabs/cardano-hw-cli/pull/112/commits/90adbcdf60606a87d5b5bed0ac75383d1334dd19

davidmisiak commented on src/crypto-providers/ledgerCryptoProvider.ts:
```diff
@@ -186,28 +182,38 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
   ): LedgerTypes.StakeCredentialParams => {
     switch (stakeCredential.type) {
       case (TxTypes.StakeCredentialType.KEY_HASH): {
         const path = findSigningPathForKeyHash(
```
:+1: Added.

davidmisiak commented on src/crypto-providers/ledgerCryptoProvider.ts:
```diff
@@ -306,6 +315,9 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
+    if (!ownersWithPath.length && signingMode === SigningMode.POOL_REGISTRATION_AS_OWNER) {
+      throw Error(Errors.MissingSigningFileForCertificateError)
+    }
     if (ownersWithPath.length > 1) throw Error(Errors.OwnerMultipleTimesInTxError)
```
Fixed.

davidmisiak commented on src/crypto-providers/ledgerCryptoProvider.ts:
```diff
@@ -511,10 +531,11 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
     const [byronWitnesses, shelleyWitnesses] = partition(
       ledgerWitnesses.map((witness) => {
```
Fixed

davidmisiak commented on src/crypto-providers/ledgerCryptoProvider.ts:
```diff
@@ -571,7 +593,7 @@ export const LedgerCryptoProvider: () => Promise<CryptoProvider> = async () => {
     const certificates = body.certificates?.map(
       (certificate) => prepareCertificate(
         certificate,
-        [...stakeSigningFiles, ...poolColdSigningFiles, ...multisigSigningFiles],
```
I believe it was a mistake.

davidmisiak commented on src/crypto-providers/signingValidation.ts:
```diff
@@ -184,21 +210,26 @@ const validateWitnessing = (params: SigningParameters): void => {
+      break
+
+    case SigningMode.PLUTUS_TRANSACTION:
+      validatePlutusSigning(body, hwSigningFileData)
```
This specifically (PLUTUS_TRANSACTION with sigining call) doesn't need to be supported, be forbidding it would bring no benefit I think. What would maybe help would be dropping support for Plutus-era txs with signing call, because that would simplify the signed tx assembling code (we decided we want to deprecate signed tx assembling only after this was implemented). But at the moment, I see no reason to remove it now - we will remove the entire signed tx assembling in a few months, and in case there are problems with Plutus-era txs with signing call, we can still point users to the witness call instead of forbidding it right now.

davidmisiak commented on src/crypto-providers/trezorCryptoProvider.ts:
```diff
@@ -207,34 +193,43 @@ const TrezorCryptoProvider: () => Promise<CryptoProvider> = async () => {
+        if (signingMode === SigningMode.PLUTUS_TRANSACTION) {
+          return { keyHash: stakeCredential.hash.toString('hex') }
+        }
+        throw Error(Errors.MissingSigningFileForCertificateError)
```
I think that would require introduction of hw-vendor-independent equivalent of `LedgerTypes.StakeCredentialParams` (path / key hash / script hash) that would be produced by the common function, and then translated into Trezor/Ledger types by hw-vendor-specific translation functions. I'm not sure this would decrease the complexity.

davidmisiak commented on src/transaction/transaction.ts:
```diff
@@ -1,45 +1,128 @@
+// If witnessSet is empty, the CBOR lib parses it as {}, otherwise it seems to always return a Map.
+// Let's convert it to Map in any case, just to make sure.
+const _parseWitnessSet = (witnessSet: unknown): WitnessSet => {
+  const clonedWitnessSet = cloneDeep(witnessSet)
```
I added the deep cloning because I don't like the idea of `TxSigned` constructor modifying the tx witness set that it received, whet it was supposed to take the parameters and create a new assembled tx.

> or generally always construct a new Map as in the last line

Without the explicit deep cloning, this would be only a shallow copy.

davidmisiak commented:
Hi @gitmachtl, I have just created a [pre-release](https://github.com/vacuumlabs/cardano-hw-cli/releases/tag/v1.10.0-rc0) with the current code from this branch. We would greatly appreciate any feedback in case you find the time to test it.

> it would be nice to have the description field copied over from the input file to the output file.

Implemented :+1: 

> So seeing it here again as type `Unwitnessed Tx AlonzoEra` is a bit confusing. Shouldn't it be `Witnessed Tx AlonzoEra` here? And, is the file at this stage also again in the CDDL format?

It should, that's a great catch, thank you! This should be fixed now. (However, I used the signed tx type `Tx AlonzoEra` and empty description instead, to be consistent with `cardano-cli sign` and the previous behavior.)

gitmachtl commented:
> Hi @gitmachtl, I have just created a [pre-release](https://github.com/vacuumlabs/cardano-hw-cli/releases/tag/v1.10.0-rc0) with the current code from this branch. We would greatly appreciate any feedback in case you find the time to test it.

hey, just saw it. will do some testing over the weekend 👍

gitmachtl commented:
@davidmisiak started to do some testing and the first one already failed.

hw-cli Prerelease 1.10-rc0
ledgerapp 4.0.0

Cardano-CLI generated CDDL formatted tx-body:
``` json
{
    "type": "Unwitnessed Tx MaryEra",
    "description": "Ledger Cddl Format",
    "cborHex": "83a500818258205a750b4f1e30a09f8cf089a38d8920aaa89518aab80d5f0d8c9100f2252e0a6a0001828258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d50d74aba4581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a44018644611223322110018c84568616c6c6f19012c43787878190190581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a1494154414441636f696e19012c581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a24378787819019042999919019082581d6052e63f22c5107ed776b70f7b92248b02552fd08f3e747bc7450994411b00000002534d4203021a0002f7ed031a0341ea3c05a1581de003d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf1b00000002534d4203a0f6"
}
```

Autocorrected to:
``` json
{
    "type": "Unwitnessed Tx MaryEra",
    "description": "Ledger Cddl Format",
    "cborHex": "83a500818258205a750b4f1e30a09f8cf089a38d8920aaa89518aab80d5f0d8c9100f2252e0a6a0001828258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d50d74aba4581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a4401864437878781901904568616c6c6f19012c4611223322110018c8581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a1494154414441636f696e19012c581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a24299991901904378787819019082581d6052e63f22c5107ed776b70f7b92248b02552fd08f3e747bc7450994411b00000002534d4203021a0002f7ed031a0341ea3c05a1581de003d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf1b00000002534d4203a0f6"
}
```

Generated witness like:
``` console
cardano-hw-cli transaction witness --tx-file /tmp/ledger-red.payment.txbody --hw-signing-file ledger-red.payment.hwsfile --hw-signing-file ledger-red.staking.hwsfile --change-output-key-file ledger-red.payment.hwsfile --change-output-key-file ledger-red.staking.hwsfile --testnet-magic 1097911063 --out-file /tmp/ledger-red.payment.txwitness
```
GUI on the LedgerWallet goes thru the Witnessing process, but after that cardano-hw-cli exited out with an error:
`The "path" argument must be of type string or an instance of Buffer or URL. Received undefined`

But is also generated a witnessFile:
``` json
{
    "type": "TxWitness MaryEra",
    "description": "",
    "cborHex": "82008258200186abff41121414f856fd8cf613399f9a30e413bb1c75c2413f46121f020349584000d095279687a0767ffe769f2117a510009575ef1a4b848cc3f719920ae780d110aa5bd22692b61af02a7b2f01c53b9d66983a023d126ee72ce689f1d6639805"
}
```

The WitnessFile is broken.

gitmachtl commented:
Next tests, again with CDDL formatted TxBody:

- Sending Lovelaces in MaryEra Format ✔️ 
- Sending Assets in MaryEra Format ✔️ 
- Sending Lovelaces in AlonzoEra Format ✔️ 
- Sending Assets in AlonzoEra Format ✔️

davidmisiak commented:
@gitmachtl Thank you very much for testing! The first transaction fails because you specified two signing files (and two signatures were indeed generated), but you specified only one witness output file, which is insufficient (each witness needs a separate output file). Could you please retry it with two output files?

In any case, the error message is terrible, I will fix it.

gitmachtl commented:
I am also having troubles with hw based pool cold keys and hw pool owners.

So, the payment.hwsfile and the staking.hwsfile are generating two witness files? I changed the method from sign to witness, i was not aware of that ok, let me try it again.

gitmachtl commented:
ok, rewards claiming is working now.

registering a stakepool with a cold-key on the ledger and/or a owner on a ledger currently results in the following errormessage when cardano-cli transaction assemble is called:

```
Command failed: transaction sign-witness  Error: Failed to decode neither the cli's serialisation format nor the ledger's CDDL serialisation format. TextEnvelope error: /dev/fd/63: TextEnvelope type error:  Expected one of: TxUnsignedByr
on, TxUnsignedShelley, TxBodyAllegra, TxBodyMary, TxBodyAlonzo Actual: Unwitnessed Tx MaryEra
TextEnvelopeCddl error: /dev/fd/63: Could not JSON decode TextEnvelopeCddl file at: /dev/fd/63 Error: Error in $: not enough input

```

gitmachtl commented:
Ok, some testing progress:

Next tests, again with CDDL formatted TxBody:

- Sending Lovelaces in MaryEra Format ✔️ 
- Sending Assets in MaryEra Format ✔️ 
- Sending Lovelaces in AlonzoEra Format ✔️ 
- Sending Assets in AlonzoEra Format ✔️ 

- Claiming Rewards in MaryEra Format ✔️ 
- Claiming Rewards in AlonzoEra Format ✔️ 

- Minting Assets on a Ledger Payment address in MaryEra Format ✔️ 
- Minting Assets on a Ledger Payment address in AlonzoEra Format ✔️ 
- Minting Assets with a HW based Policy on a Ledger Payment address in MaryEra Format ✔️ 
- Minting Assets with a HW based Policy on a Ledger Payment address in AlonzoEra Format ✔️ 

- Burning Assets on a Ledger Payment address in MaryEra Format ✔️ 
- Burning Assets on a Ledger Payment address in AlonzoEra Format ✔️ 
- Burning Assets with a HW based Policy on a Ledger Payment address in MaryEra Format ✔️ 
- Burning Assets with a HW based Policy on a Ledger Payment address in AlonzoEra Format ✔️ 

- Stakepool registration with HW based Cold keys on the Ledger ✔️ 
- Stakepool registration with a Ledger owner ✔️ 

... to be continued

gitmachtl commented:
@davidmisiak question, we had the restriction before that a stakepool with cold keys on the hw wallet can only be retired via a signed transaction from a payment wallet also on the same hw wallet. is this limitation still in place now with all transactions switching to witness+assemble?

gitmachtl commented:
- Witness a Trezor PoolOwner in a PoolRegistration :x:

This actions fails with: `Missing signing file for certificate`

davidmisiak commented:
> How can the trezor witness file be in Alonzo Era format while the TxBody as an input is in Mary Era format?

This seems suspicious - hw-cli reads the era from the tx body envelope type and writes it to the witness file envelope type. Are you sure the `/tmp/trezor.payment.txwitness-staking` path really points to the witness which was generated for the Mary era tx body?

Similarly, the other errors might be caused by incorrect signing file paths / witness file paths when assembling. Is there a chance there were some stale files from previous signing sessions / Ledger files instead? I'm just trying to rule out the possible reasons that come to my mind; if the errors persist, I will investigate them further.

gitmachtl commented:
Running

``` console
cardano-hw-cli transaction witness --tx-file /tmp/trezor.payment.txbody --hw-signing-file trezor.payment.hwsfile --hw-signing-file trezor.staking.hwsfile --change-output-key-file trezor.payment.hwsfile --change-output-key-file trezor.staking.hwsfile --testnet-magic 1097911063 --out-file /tmp/trezor.payment.txwitness-payment --out-file /tmp/trezor.payment.txwitness-staking
```

Only returns one witness file, but not two! The 2nd WitnessFile `/tmp/trezor.payment.txwitness-staking` is missing.

davidmisiak commented:
> So the same witness call with a StakeKeyDeregistration returns two witnessfiles, but a StakeKeyRegistration only one.

That's really weird, trezorCryptoProvider treats key registration and deregistration the same way. One more idea - the `transaction submit Error` above actually complaints about a UTXO being witnessed incorrectly - maybe the generated witness is actually the staking witness, and the payment witness is missing? Could you perhaps provide the `/tmp/trezor.payment.txwitness-payment` file contents please? Also, the deregistration tx file (that was successful) could help as well.

gitmachtl commented:
Here are the files for the **successful stakeKey deregistration**:

/tmp/trezor.payment.txbody: 
``` json
{
    "type": "Unwitnessed Tx MaryEra",
    "description": "Ledger Cddl Format",
    "cborHex": "83a5008182582062592cc750f8833f791a514ae4a07fbc484ef206271389cf6aac2362924b268200018182583900553eef4a0e12ff6a0729d8009a6c89b03d4e206561d631e493654e8aa6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148a1a0090aadd021a0002bc75031a034236b8048182018200581ca6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148aa0f6"
}
```

/tmp/trezor.payment.txwitness-payment:
``` json
{
    "type": "TxWitness MaryEra",
    "description": "",
    "cborHex": "8200825820433e2661aded5f3c331a5a31a192b02ee05c761f97d0fbb9fcd8d37f200c5b8a58403608fc60f221c942493abaebbcfbd50f11c9e0785e15bb329558e5852617fbea0db2723424542a8bfe5d7af14a2b56b2167321e02d424e660d1c1f77ad874c0b"
}
```

/tmp/trezor.payment.txwitness-staking:
``` json
{
    "type": "TxWitness MaryEra",
    "description": "",
    "cborHex": "8200825820e219d6d63d7a1aac472d457a3299e76f5e6ef8c38897411a32dc8aefdc45144c58408db068539ced37358ca799bfd652d47fc730ac7e7391b91be3c7691f73df044d31d0ab25630da3070cd31703e8ea1d81150ccefaa5efd54db5f7d62f8cdaaf07"
}
```

/tmp/trezor.payment.tx
``` json
{
    "type": "Witnessed Tx MaryEra",
    "description": "Ledger Cddl Format",
    "cborHex": "83a5008182582062592cc750f8833f791a514ae4a07fbc484ef206271389cf6aac2362924b268200018182583900553eef4a0e12ff6a0729d8009a6c89b03d4e206561d631e493654e8aa6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148a1a0090aadd021a0002bc75031a034236b8048182018200581ca6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148aa10082825820433e2661aded5f3c331a5a31a192b02ee05c761f97d0fbb9fcd8d37f200c5b8a58403608fc60f221c942493abaebbcfbd50f11c9e0785e15bb329558e5852617fbea0db2723424542a8bfe5d7af14a2b56b2167321e02d424e660d1c1f77ad874c0b825820e219d6d63d7a1aac472d457a3299e76f5e6ef8c38897411a32dc8aefdc45144c58408db068539ced37358ca799bfd652d47fc730ac7e7391b91be3c7691f73df044d31d0ab25630da3070cd31703e8ea1d81150ccefaa5efd54db5f7d62f8cdaaf07f6"
}
```

-----

And here are the files for a **failing stakeKey registration**:

/tmp/trezor.payment.txbody
``` json
{
    "type": "Unwitnessed Tx MaryEra",
    "description": "Ledger Cddl Format",
    "cborHex": "83a50081825820c4b93b9f151a07890c65bc674a77c0e48dd2cd027d5e83ba6041eb26bfd348a700018182583900553eef4a0e12ff6a0729d8009a6c89b03d4e206561d631e493654e8aa6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148a1a006f69bc021a0002bca1031a0342e89c048182008200581ca6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148aa0f6"
}
```

/tmp/trezor.payment.txwitness-payment
``` json
{
    "type": "TxWitness MaryEra",
    "description": "",
    "cborHex": "8200825820433e2661aded5f3c331a5a31a192b02ee05c761f97d0fbb9fcd8d37f200c5b8a584050d053aabc7445047de61570e3a340d78bfcff8c095f49abe3d7b82552ea73111d2fbb15421a427ec9d1f75deead263853ecd2032f2ed5a877b85ba8d424ca09"
}
```

/tmp/trezor.payment.txwitness-staking-> missing

``` console
$ cardano-hw-cli device version
Trezor app version 2.4.4
$ cardano-hw-cli version
Cardano HW CLI Tool version 1.10.0-rc0
Commit hash: 9c0621df4d60e02602eb20ceb9e78b2c3b5a4319
```

davidmisiak commented:
Thanks, but I'm afraid that both tx files and payment witnesses are actually the same (the deregistration ones).

gitmachtl commented:
Let me recheck it, maybe i made a copy paste error

gitmachtl commented:
I tried it again and updated the files again in the post above.

gitmachtl commented:
here are the files in comparision using a ledger nano. 

**stakeKey deregistration**

/tmp/ledger-red.payment.txbody
``` json
{
    "type": "Unwitnessed Tx AlonzoEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a5008182582037dc60aceb9846be934f3789d3334918d7d9937e4ea948ff606c33994f3a50f40001818258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d4c1d2cfa4581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a4401864437878781901904568616c6c6f19012c4611223322110018b4581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a24873636f6f70209a0a0a494154414441636f696e19012c581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a24299991901904378787819019a021a0002e875031a034302be048182018200581c03d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cfa0f5f6"
}
```

/tmp/ledger-red.payment.txwitness-payment
``` json
{
    "type": "TxWitness AlonzoEra",
    "description": "",
    "cborHex": "82008258200186abff41121414f856fd8cf613399f9a30e413bb1c75c2413f46121f020349584004ffe0cd30921758e5c2d9b39beebf3054a6c452cf08ed5afc0e370dc7831e00e7d188f6848eaa04e90b3c0740334676165b1db3268a38ee1659856fd3224808"
}
```

/tmp/ledger-red.payment.txwitness-staking
``` json
{
    "type": "TxWitness AlonzoEra",
    "description": "",
    "cborHex": "8200825820c13745be35c2dfc3fa9523140030dda5b5346634e405662b1aae5c61389c55b35840cf9cc94bcb217adfb518ac8487dcc916afd74255f2f31a40090087f0a03078c95101ffeb43302ca548f4a4b7531c94cbb86422744a792e90d354d6d7e0c0b10c"
}
```

/tmp/ledger-red.payment.tx
``` json
{
    "type": "Witnessed Tx AlonzoEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a5008182582037dc60aceb9846be934f3789d3334918d7d9937e4ea948ff606c33994f3a50f40001818258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d4c1d2cfa4581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a4401864437878781901904568616c6c6f19012c4611223322110018b4581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a24873636f6f70209a0a0a494154414441636f696e19012c581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a24299991901904378787819019a021a0002e875031a034302be048182018200581c03d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cfa10082825820c13745be35c2dfc3fa9523140030dda5b5346634e405662b1aae5c61389c55b35840cf9cc94bcb217adfb518ac8487dcc916afd74255f2f31a40090087f0a03078c95101ffeb43302ca548f4a4b7531c94cbb86422744a792e90d354d6d7e0c0b10c8258200186abff41121414f856fd8cf613399f9a30e413bb1c75c2413f46121f020349584004ffe0cd30921758e5c2d9b39beebf3054a6c452cf08ed5afc0e370dc7831e00e7d188f6848eaa04e90b3c0740334676165b1db3268a38ee1659856fd3224808f5f6"
}
```

-----

**stakeKey registration**

/tmp/ledger-red.payment.txbody
``` json
{
    "type": "Unwitnessed Tx AlonzoEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a5008182582026cbf637e84f47bac3349383c3d2cce778424cb70d2d26b2684b486712137fda0001818258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d4a065aea4581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a4401864437878781901904568616c6c6f19012c4611223322110018b4581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a24873636f6f70209a0a0a494154414441636f696e19012c581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a24299991901904378787819019a021a0002e8a1031a03430368048182008200581c03d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cfa0f5f6"
}
```

/tmp/ledger-red.payment.txwitness-payment
``` json
{
    "type": "TxWitness AlonzoEra",
    "description": "",
    "cborHex": "82008258200186abff41121414f856fd8cf613399f9a30e413bb1c75c2413f46121f0203495840cde0135a4de0caf23aa1f2efa5faffc5c3a3d08b79038cf5ef5fcc5a2ee362b1254861778b84a709b18835866dbe9f887562a5160ab0f1978d06d2500d1e0203"
}
```

/tmp/ledger-red.payment.txwitness-staking
``` json
{
    "type": "TxWitness AlonzoEra",
    "description": "",
    "cborHex": "8200825820c13745be35c2dfc3fa9523140030dda5b5346634e405662b1aae5c61389c55b35840ee382a41598c56bec913b2ba3a3cc4baca46a7247107f6bf96d5a81a2fc64a27728090c5abd05e3bb661147031239cbc96493e5f286969187cb9bb466b3c9d04"
}
``` 

/tmp/ledger-red.payment.tx
``` json
{
    "type": "Witnessed Tx AlonzoEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a5008182582026cbf637e84f47bac3349383c3d2cce778424cb70d2d26b2684b486712137fda0001818258390074976c54afaf444f7cd499bd8519aac6592b13b22b9d5817f0da5c5203d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf821b000000e8d4a065aea4581c199062e26a0ea1370249e71e6224c6541e7825a192fe42c57aa538c3a2581c41616461476f6c64656e526566657272616c3136333930343938363601581c4161646153696c766572526566657272616c3136343036313131353401581c2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15a4401864437878781901904568616c6c6f19012c4611223322110018b4581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a24873636f6f70209a0a0a494154414441636f696e19012c581c6b79ff1505414bd7ebba7a9b3242ba02810f0b91ea9e391c6e0b7409a24299991901904378787819019a021a0002e8a1031a03430368048182008200581c03d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cfa10082825820c13745be35c2dfc3fa9523140030dda5b5346634e405662b1aae5c61389c55b35840ee382a41598c56bec913b2ba3a3cc4baca46a7247107f6bf96d5a81a2fc64a27728090c5abd05e3bb661147031239cbc96493e5f286969187cb9bb466b3c9d048258200186abff41121414f856fd8cf613399f9a30e413bb1c75c2413f46121f0203495840cde0135a4de0caf23aa1f2efa5faffc5c3a3d08b79038cf5ef5fcc5a2ee362b1254861778b84a709b18835866dbe9f887562a5160ab0f1978d06d2500d1e0203f5f6"
}
```

works like a charm, no change in the scripts, no change in the callings

gitmachtl commented:
Maybe there is something strange going on with the autocorrection of the cannonical order?

This is the txbody before the autocorrection:
``` json
{
    "type": "Unwitnessed Tx AlonzoEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a70081825820a34c02ccca70728ecea66a41bb0307159b73ef543dece433069d0275ab7b1492000d80018182583900553eef4a0e12ff6a0729d8009a6c89b03d4e206561d631e493654e8aa6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148a821a003d3beaa1581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a1494154414441636f696e05021a0002c4e1031a034308bd048182008200581ca6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148a0e80a0f5f6"
}
```

And this is the txbody after the autocorrection:
``` json
{
    "type": "Unwitnessed Tx AlonzoEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a50081825820a34c02ccca70728ecea66a41bb0307159b73ef543dece433069d0275ab7b149200018182583900553eef4a0e12ff6a0729d8009a6c89b03d4e206561d631e493654e8aa6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148a821a003d3beaa1581c34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518a1494154414441636f696e05021a0002c4e1031a034308bd048182008200581ca6b791a81822aa986e38a8270f2917d9aa93bd3577aad828b37c148aa0f5f6"
}
```

Why is it shorter?

Also, cardano-hw-cli is crashing when calling it with the not corrected txBody file.

davidmisiak commented:
Alright, I figured it out. Cardano stake key registration doesn't actually require a witness, see e.g. section 3.3.3 [here](https://hydra.iohk.io/build/6752481/download/1/delegation_design_spec.pdf):  "We do not require a witness to register a stake address (besides, of course, any witnesses needed for the transaction that is used to post the certificate)."

Trezor's behavior is therefore the correct one - Trezor Connect uses the staking path when displaying the certificate on screen, but doesn't generate the signature. On the other hand, Ledger doesn't distinguish between certificates in this regard and always generates the signature. I believe this inconsistency should be easily fixable; I'll notify you when it's done (hopefully before the 1.10 release). In the meantime, there are several ways how to tackle this - either using the `cardano-hw-cli sign` call temporarily (which hides the inconsistency), or treating Trezor and Ledger differently when signing a staning key registration.

I'll also add a warning message when there are superfluous witness out files specified on the command line.

Are there any other issues? Did you manage to resolve the `Witness a Trezor PoolOwner in a PoolRegistration` one? In case not, could you please provide the relevant files and commands please?

gitmachtl commented:
Had no time to revisite the PoolOwner one yet. But nothing changed, so the problem is still there.

Ledger does not need a witness to register a stake address? So they don't need the staking witness i guess, because the registration must be paid from an address on the same ledger. And so this is enough for them?

davidmisiak commented:
> Why is it shorter?

That is a part of autocorrection - not only fixing item order, but also removing empty optional arrays (e.g. collaterals, when there are none).

> Also, cardano-hw-cli is crashing when calling it with the not corrected txBody file.

This is expected -  HW wallets use the canonical serialization when creating the signatures, which means that the signatures would not match the original tx body CBOR. But before exiting, hw-cli should print a message about the tx body not being canonically formatted - is this the case?

gitmachtl commented:
> This is expected - HW wallets use the canonical serialization when creating the signatures, which means that the signatures would not match the original tx body CBOR. But before exiting, hw-cli should print a message about the tx body not being canonically formatted - is this the case?

Yes it is, but i am wondering. This is a payment address with only one utxo that contains some ada an a single kind of assets. so even that txbody output from the cardano-cli is not in canonical order? strange.

gitmachtl commented:
Wow ok, i tried to only use one witness with the trezor device for the stakeKey registration and it went thru. This is unexpected.

I went on and tried to register a normal cli based staking address also without providing the skey for it, and it worked. So we "most of us" did is wrong the whole time. This is surprising. 😱

gitmachtl commented:
So, after your correction i guess the output of the cardano-hw-cli transaction witness call for a stakeKey registration will also only output the witness for the payment. Right?

davidmisiak commented:
> So, after your correction i guess the output of the cardano-hw-cli transaction witness call for a stakeKey registration will also only output the witness for the payment. Right?

I think so, yes.

> The message does not appear, it exits with errorcode 1, or the message is not written to stderr and stdout instead.

Exit code 1 is correct. I don't know what could be the reason of error message missing - I tried both signing and witnessing both tx-file and tx-body-file (before transformation) and this message was always printed to stderr: `Error: Transaction CBOR contains fixable errors, please run "transform" or "transform-raw" command first`.

> so even that txbody output from the cardano-cli is not in canonical order? strange.

The order is canonical, but cardano-cli includes an empty collaterals array, unfortunately. And if the tx contains no collateral inputs, [CIP-0021](https://cips.cardano.org/cips/cip21/) requires the collaterals array to be omitted entirely (which is done by `cardano-hw-cli transform`).

gitmachtl commented:
Ok, so if we don't need an stake.skey to do a stakeKey registration. I guess the stake.hwsfile is than only needed to show the information about the stakeaddress on the hw-wallet GUI?

davidmisiak commented:
@gitmachtl I tried to witness a pool registration as owner with Trezor, but I didn't encounter any `Missing signing file for certificate` error (a staking witness was generated instead, as expected). Did you use the same stake signing file that was referenced in the pool owner tx item? If so, can you please provide the tx file and the exact hw-cli command?

gitmachtl commented:
> @gitmachtl I tried to witness a pool registration as owner with Trezor, but I didn't encounter any `Missing signing file for certificate` error (a staking witness was generated instead, as expected). Did you use the same stake signing file that was referenced in the pool owner tx item? If so, can you please provide the tx file and the exact hw-cli command?

ok, i think there was a bug with the keys when i installed various fw version on the trezor. so it created new internal keys.

I tried various combinations now with MultiPoolOwner-Stakepool-Registrations. A mix of Ledger, Trezor devices and CLI keys. Also Ledger, Trezor, CLI keys as rewards account. I also tried to use the hw-based cold keys together with a ledger owner and a trezor owner in a poolRegistration. All worked well now. I am running Fw 2.4.4 on the Trezor. But i don't think that this made the difference. All transactions in CDDL format and AlonzoEra. 👍

davidmisiak commented:
Great, I'm glad everything works. And once again thank you very much for your effort, we really appreciate it! In case any new issues appear, please let us know.

> Creating Cold Pool Keys on the Trezor is still not possible, do you know if this is on the list in the future?

This is not planned in the immediate future as far as I know.

> When using the hw-based policyKeys for example to mint/burn Assets. You can witness an unwitnessed txBody with the wrong hw-wallet. cardano-hw-cli does not show this as the wrong key/certificate. You can also assemble it, but only on the submit an error appears. Any good way to catch this?

I don't think this is possible in general. A transaction file doesn't necessarily contain the info required to determine whether a minting key is a correct one - the native script - at the moment of witnessing (it may be added later). As you've mentioned, even cardano-cli let's you assemble almost anything, and checks for witness completeness and correctness only when submitting the assembled tx, so I doubt we can do anything beyond that.

davidmisiak commented:
@gitmachtl FYI, we've added the witness count vs. out files count validation and updated Ldegerjs to no longer generate key registration certificate staking witnesses. There is also a 1.10.0-rc1 pre-release with these changes.

> we had the restriction before that a stakepool with cold keys on the hw wallet can only be retired via a signed transaction from a payment wallet also on the same hw wallet. is this limitation still in place now with all transactions switching to witness+assemble?

If you mean the fact that `hw-cli transaction sign` complains about a missing payment signing file, that's because the `sign` call is supposed to generate a signed transaction ready to be submitted (which obviously requires at least one payment signing file). On the other hand, `hw-cli transaction witness` is able to generate the pool cold key witness separately (and it has been this way in the previous release as well), so you can create the payment witness for the retirement tx e.g. from a non-hw wallet and then assemble the witnesses. Note that if you want to use a hw wallet to create the payment witness, too, the same hw wallet indeed needs to control the pool cold key, because you need to provide the pool cold key signing file when creating _any_ hw witness (because the pool cold key path has to be displayed on the hw screen). This has nothing to do with the signing vs. witnessing flow in hw-cli, it's how HW wallets work and nothing has changed in this regard.

davidmisiak commented:
@gitmachtl FYI, we've added the witness count vs. out files count validation and updated Ldegerjs to no longer generate key registration certificate staking witnesses. There is also a [1.10.0-rc1 pre-release](https://github.com/vacuumlabs/cardano-hw-cli/releases/tag/v1.10.0-rc1) with these changes.

> we had the restriction before that a stakepool with cold keys on the hw wallet can only be retired via a signed transaction from a payment wallet also on the same hw wallet. is this limitation still in place now with all transactions switching to witness+assemble?

If you mean the fact that `hw-cli transaction sign` complains about a missing payment signing file, that's because the `sign` call is supposed to generate a signed transaction ready to be submitted (which obviously requires at least one payment signing file). On the other hand, `hw-cli transaction witness` is able to generate the pool cold key witness separately (and it has been this way in the previous release as well), so you can create the payment witness for the retirement tx e.g. from a non-hw wallet and then assemble the witnesses. Note that if you want to use a hw wallet to create the payment witness, too, the same hw wallet indeed needs to control the pool cold key, because you need to provide the pool cold key signing file when creating _any_ hw witness (because the pool cold key path has to be displayed on the hw screen). This has nothing to do with the signing vs. witnessing flow in hw-cli, it's how HW wallets work and nothing has changed in this regard.

gitmachtl commented:
> There is also a [1.10.0-rc1 pre-release](https://github.com/vacuumlabs/cardano-hw-cli/releases/tag/v1.10.0-rc1) with these changes.

Thx, i have retried it with 1.10.0-rc1 and also removed the staking witness in my scripts for the stakeKey registration. It works not for Ledger and Trezor. 👍

davidmisiak commented:
Non-beta release here: https://github.com/vacuumlabs/cardano-hw-cli/releases/tag/v1.10.0

Thank you once again for cooperation!
